rfd/00XY-session-locking.md

```diff
@@ -0,0 +1,262 @@
+// wildcards of regular expressions.
+message SessionLockTarget {
+    // User specifies the name of a Teleport user.
+    string User;
+    // Role specifies the name of a RBAC role.
+    string Role;
+    // Cluster specifies the name of a Teleport cluster.
+    string Cluster;
+    // Login specifies the name of a local UNIX user.
+    string Login;
+    // Node specified the name or UUID of a node.
+    string Node;
+}
+```
+
+In most cases `SessionLockTarget` should list just a single attribute.
+(It would be preferable to define it as a `oneof`, however setting a custom
+`jsontag` for `oneof` is [not supported by `gogoproto`](https://github.com/gogo/protobuf/issues/623).)
+
+Note that `SessionLock` is not a singleton resource: when there are multiple
+`SessionLock`s stored (and in force), it suffices for a session to be matched
+by any one of the locks to be terminated/disabled.
+
+If the conditions encoded by a set of `SessionLock`s was to be expressed in
+a single logical formula, the conditions within `SessionLockTarget` would be
+connected with logical conjunction while those of multiple `SessionLock`s would
+be connected with logical disjunction.  This is similar to the notion of
+_disjunctive normal form_ of propositional logic.  In the ordinary case where
+`SessionLockTarget` lists just a single attribute, the formula would reduce to
+a simple disjunction of atomic constraints.
+
+#### Expiration & audit
+
+There is no explicit expiry field in `SessionLockSpecV2`.  The common `Metadata.Expires`
+field is used instead.
+
+Since the `Metadata.Expires` field is also used by the backends to delete stale
+data, this automatically guarantees no expired `SessionLocks` will be returned
+by methods like `GetSessionLocks`, which in turn implies only unexpired locks
+can ever be presented to the user in summary outputs.
+
+Historical `SessionLock` records can be reconstructed only from the audit log.
+Every `SessionLock` create/update/delete should cause an audit event to be
+emitted.
+
+#### Relation to `User.Status`
+
+There already exists a field of `User` resource that is used to
+capture a user lock in connection with failed Web UI login attempts.
+
+This `Status` field and its `LoginStatus` definition are superseded by
+`SessionLock`.  All of its use cases should be converted to `SessionLock`.
+
+The `SessionLock` approach allows to specify locks for entities that are only
+yet to exist or exist merely transiently (such as SSO user objects).  It could
+also help with alleviating the load associated with caching `User` resources.
+
+#### `tctl` support
+
+`SessionLock` resources can be managed using `tctl [get|create|rm]`.  In this
+way, it is possible to update (e.g. delay) or remove a session lock after it
+has been created.
+
+There will be a special `tctl sessions lock` helper provided, to facilitate
+supplying time information when creating new `SessionLock`s, see Scenarios below.
```

Comment:
`tctl [get|create|rm]` is virtually for free once the resource type is defined.

Reply:
My thinking here was that since locks aren't a static configuration object, we don't *need* to expose their internal structure.  If we don't expose their internal structure, that gives us more freedom to make breaking changes in the future.  Its always more risky to change resources that are exposed via the `tctl [get|create|rm]` commands since we don't control all the code that builds the resources (i.e. customers may have scripts that build the resources).  Its a fairly minor and hypothetical concern, so feel free to ignore it :)